### PR TITLE
[FEATURE] QTI import: Add XML preprocess step to set missing title

### PR DESCRIPTION
**Pull request by ttton**

### CODE REVIEW

File: src/Processors/QtiV2/In/ItemMapper.php [hunk 1]
```diff
@@ -22,25 +24,56 @@ public function __construct(ItemBuilderFactory $itemBuilderFactory)
...
         LogService::flush();
 
-        $xmlDocument = new XmlDocument();
-        if ($validate === false) {
-            LogService::log('QTI pre-validation is turned off, some invalid attributes might be stripped from XML content upon conversion');
-        }
-        $xmlDocument->loadFromString($xmlString, $validate);
+        $xmlPreProcessings = [
+            AppContainer::getApplicationContainer()->get('xml_assessment_items_processing'),
+        ];
```
Comment (ekcolysp) [inline src/Processors/QtiV2/In/ItemMapper.php]: put this in another method.

File: src/Processors/QtiV2/In/ItemMapper.php [hunk 2]
```diff
@@ -22,25 +24,56 @@ public function __construct(ItemBuilderFactory $itemBuilderFactory)
...
+
+        $xmlString = $this->preProcessXml($xmlString, $xmlPreProcessings);
```
Comment (ekcolysp) [inline src/Processors/QtiV2/In/ItemMapper.php]: change preProcessXml to processXml

File: src/Processors/QtiV2/In/ItemMapper.php [hunk 3]
```diff
@@ -51,16 +84,37 @@ public function parseWithAssessmentItemComponent(AssessmentItem $assessmentItem)
...
-            AppContainer::getApplicationContainer()->get('identifiers_processing')
+            AppContainer::getApplicationContainer()->get('identifiers_processing'),
         ];
 
-        // Pre-processing works
-        /** @var ProcessingInterface $processing */
-        foreach ($processings as $processing) {
-            $assessmentItem = $processing->processAssessmentItem($assessmentItem);
-        }
+        $assessmentItem = $this->preProcessQtiAssessmentItem($assessmentItem, $processings);
```
Comment (ekcolysp) [inline src/Processors/QtiV2/In/ItemMapper.php]: change this to `processing...`

File: src/Processors/QtiV2/In/ItemMapper.php [hunk 4]
```diff
@@ -51,16 +84,37 @@ public function parseWithAssessmentItemComponent(AssessmentItem $assessmentItem)
...
+
+        $assessmentItem = $this->validateQtiAssessmentItem($assessmentItem);
 
-        $assessmentItem = $this->validateAssessmentItem($assessmentItem);
+        // Conversion from QTI item to Learnosity item and questions
+        list($item, $questions) = $this->buildLearnosityItemFromQtiAssessmentItem($assessmentItem);
+
+        list($item, $questions) = $this->postProcessLearnosityItem($item, $questions, $processings);
```
Comment (ekcolysp) [inline src/Processors/QtiV2/In/ItemMapper.php]: change this to `processing...`

File: src/Processors/QtiV2/In/Processings/QtiV2AssessmentItemProcessing.php [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+     */
+    protected function handleAssessmentItemInvalidTitle(SimpleXMLElement $xmlElement)
+    {
+        if ($this->isXmlElementAssessmentItem($xmlElement)) {
+            $nodeAttributes = $xmlElement->attributes();
+
+            if (!isset($nodeAttributes['title'])) {
+                $xmlElement->addAttribute('title', $nodeAttributes['identifier']);
+            } elseif (empty($nodeAttributes['title'])) {
+                $nodeAttributes['title'] = (string) $nodeAttributes['identifier'];
```
Comment (cameron-simpson) [inline src/Processors/QtiV2/In/Processings/QtiV2AssessmentItemProcessing.php]: Wondering if the code would be less confusing if we went:

  $titleAttribute = (string) $nodeAttributes['identifier'];

and pass it in both branches of the "if".

**Pull request merged**